In [20]:
from __future__ import print_function, division
import os
import sys
import numpy as np
import pylab as pl


%pylab inline

from IPython.display import Image
#Image(filename='../plotsforclasses/NYCReentryprogram_title.png')

Populating the interactive namespace from numpy and matplotlib


In [21]:
def evalChisq(values):
    '''Evaluates the chi sq from a contingency value
    Arguments:
    values: 2x2 array or list, the contingengy table
    '''
    if not (len(values.shape) == 2 and values.shape == (2,2)):
        print ("must pass a 2D array")
        return -1
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()

$H_0: P_0 - P_1 \geq$ 0
<br />
$H_a: P_0 - P_1 $< 0 

In [23]:
# first thing: choose your significance threshold
alpha=0.05

# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes
n_0 = 409
n_1 = 564
    
#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

we must assess the statistical significance


In [24]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.421047276465


In [25]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

In [26]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)

20.769786540820895


In [27]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9984

def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


In [29]:
Ntot = 973 # a + b + c + d = tot

sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])

print (evalChisq(sample_values))

436.22346257536003


Since our chi squared value is 436 and we are looking for a p-value less than 0.05, we must reject the null hypothesis. This is because the chi-squared value we should be expecting at this p-value should be 3.84. If it were less, that would be fine, but since it is larger, we reject the null.

# For me to do:

$H_0: P_0 - P_1 \geq$ 0
<br />
$H_a: P_0 - P_1 $< 0

In [31]:
# choosing your significance threshold
alpha=0.05

# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 11.7 * 0.01 
P_1 = 10 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds... but let's see what happens anyway :)")
else:
    print ("we must assess the statistical significance")

#sample sizes
n_0 = 409
n_1 = 568
    
#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

the Null holds... but let's see what happens anyway :)


In [32]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.107116683726


In [33]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

In [38]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)

-0.8476402052199316


In [40]:
## p-value for being convicted of a felony: 
## Looking for a value that corresponds with absolute value of -0.8466, rounding to 0.8, and alpha = 0.05. 
#We get 0.8023

p_2y = 1 - 0.8023

def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


|convicted of a fellony|     yes     | no         |                
|----------------------|-------------|------------|----------------|
| test sample          | 0.1 * 568   |0.9 * 568   |   568          |
| control sample       |0.117 * 409  |0.883 * 409 |   409          |
|                      |             |            |                |
| total                |  104.25     | 868.75     |   977          |

In [39]:
#Now for Chi-Squared evaluation
Ntot = 977 # a + b + c + d = tot

sample_values = np.array([[0.1 * 568, 0.9 * 568], [0.117 * 409, 0.883 * 409]])

print (evalChisq(sample_values))

0.7184939175052887


This chi-squared value lies between 0.5 and 0.25 probability of this event occuring. These probabilities are much 
larger than the 0.05 probability that being convicted of a felony between the control and test groups were correlated.
We already rejected the null based off the z-test. This chi squared evaluation only further proves that there is no
relationship between these programs and recidivism.